src/log_file.ml

```diff
@@ -0,0 +1,207 @@
+       the entries within each bucket being relatively out of order). *)
+    let unneeded_bits = Key.hash_size - t.bucket_count_log2 in
+    (Key.hash key lsr unneeded_bits) land ((1 lsl t.bucket_count_log2) - 1)
+
+  let resize t =
+    (* Scale the number of hashset buckets. *)
+    t.bucket_count_log2 <- t.bucket_count_log2 + 1;
+    let new_bucket_count = 1 lsl t.bucket_count_log2 in
+    let new_hashset = Array.make new_bucket_count Small_list.empty in
+    ArrayLabels.iteri t.hashset ~f:(fun i bucket ->
+        (* The bindings in this bucket will be split into two new buckets, using
+           the next bit of [Key.hash] as a discriminator. *)
+        let bucket_2i, bucket_2i_plus_1 =
+          Small_list.to_list bucket
+          |> List.partition (fun offset ->
+                 let key = key_of_offset t offset in
+                 let new_index = elt_index t key in
+                 assert (new_index lsr 1 = i);
+                 new_index land 1 = 0)
+        in
+        new_hashset.(2 * i) <- Small_list.of_list bucket_2i;
+        new_hashset.((2 * i) + 1) <- Small_list.of_list bucket_2i_plus_1);
+    t.hashset <- new_hashset
+
+  (** Replace implementation that only updates in-memory state (and doesn't
+      write the binding to disk). *)
+  let replace_memory t key offset =
+    if t.cardinal > 2 * Array.length t.hashset then resize t;
+    let elt_idx = elt_index t key in
+    let bucket = t.hashset.(elt_idx) in
+    let bucket =
+      let key_found = ref false in
+      let bucket' =
+        Small_list.map bucket ~f:(fun offset' ->
+            if !key_found then
+              (* We ensure there's at most one binding for a given key *)
+              offset'
+            else
+              let key' = key_of_offset t offset' in
+              match Key.equal key key' with
+              | false -> offset'
+              | true ->
+                  (* New binding for this key *)
+                  key_found := true;
+                  offset)
+      in
+      match !key_found with
+      | true ->
+          (* We're replacing an existing value. No need to change [cardinal]. *)
+          bucket'
+      | false ->
+          (* The existing bucket doesn't contain this key. *)
+          t.cardinal <- t.cardinal + 1;
+          Small_list.cons offset bucket
+    in
+    t.hashset.(elt_idx) <- bucket
+
+  let replace t key value =
+    let offset = IO.offset t.io in
+    Entry.encode' key value t.append_io;
+    replace_memory t key offset
+
+  let sync_entries ~min t =
+    IO.iter_keys ~min (fun offset key -> replace_memory t key offset) t.io
+
+  let reload t =
+    clear_memory t;
+    sync_entries ~min:Int63.zero t
+
+  let create io =
+    let cardinal = Int63.(to_int_exn (IO.offset io / Entry.encoded_sizeL)) in
+    let bucket_count_log2, bucket_count =
+      let rec aux n_log2 n =
+        if n >= cardinal then (n_log2, n)
+        else if n * 2 > Sys.max_array_length then (n_log2, n)
```

Comment:
since we are guarding against this case here, maybe we should do it when resizing too? (though it seems likely to reach such large arrays)